Implement UI components for executor secrets

eseliger opened the pull request:
## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.07% (+2.06 kb) | 0.19% (+27.47 kb) 🔺 | 0.21% (+25.42 kb) 🔺 | 0.41% (+3) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/es/executor-secrets-ui/compare-report.html;tab=live_object) for a full comparison between the commits 0138f35 and fc48070 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

eseliger commented on client/web/src/enterprise/user/settings/sidebaritems.ts:
```diff
@@ -22,6 +22,11 @@ export const enterpriseUserSettingsSideBarItems: UserSettingsSidebarItems = [
+    {
+        to: '/executors/secrets',
+        label: 'Executor secrets',
+        condition: ({ user: { viewerCanAdminister } }) => viewerCanAdminister,
```
Question: Should this menu item be hidden when executors are not configured? 
That could leave a weird state where they were enabled in the past and now you cannot access and delete your existing secrets anymore. Hence, I'm leaning towards always showing it. What do people think?

adeola-ak replied:
when I used `areExecutorsEnabled` in the past if they were ever enabled even if they are no longer enabled, it will still be true. Can you use that? I think it makes sense to hide it for the case that they are not enabled

courier-new replied:
In general our approach to UI seems to be that it's better not to hide things, just to disable them. Since there isn't a concept of disabled sidebar items, maybe instead it would make sense to show a banner on the secrets list page (or a different version of the list?) if executors aren't enabled.

(If `areExecutorsConfigured` is `true` even after executors are disabled, that doesn't seem good. 😅)

courier-new commented on client/web/src/enterprise/executors/secrets/ExecutorSecretNode.tsx:
```diff
@@ -0,0 +1,132 @@
+        refetchAll()
+    }, [refetchAll, buttonReference])
+
+    const headingAriaLabel = 'Secret value'
```
I was kind of confused by this label because it made me think it was describing an element with the literal _value_ of the _secret_, even though we just show the key and metadata about it. Maybe just "Secret" would be better? Or even omitting it entirely?

courier-new commented on client/web/src/enterprise/executors/secrets/ExecutorSecretsListPage.tsx:
```diff
@@ -0,0 +1,211 @@
+                description={headerLine}
+                actions={
+                    <>
+                        <Button onClick={onClickAdd} aria-label="Add new secret value" variant="primary">
```
I think the button text is sufficient as a label here.

```suggestion
                        <Button onClick={onClickAdd} variant="primary">
```

courier-new commented on client/web/src/enterprise/executors/secrets/ExecutorSecretsListPage.tsx:
```diff
@@ -0,0 +1,211 @@
+            )}
+
+            <div className="d-flex mb-3">
+                {Object.values(ExecutorSecretScope).map(scope => (
+                    <ExecutorSecretScopeSelector
+                        key={scope}
+                        scope={scope}
+                        label={executorSecretScopeContext(scope).label}
+                        onSelect={() => setSelectedScope(scope)}
+                        selected={scope === selectedScope}
+                        description={executorSecretScopeContext(scope).description}
+                    />
+                ))}
```
I got really confused by this when I first saw it in the Storybook story, especailly since right now the only scope is "Batch Changes", which results in a disabled button. I think this probably ought to be a dropdown, not buttons, and there should probably be a label for "Scope" or something next to.

eseliger replied:
Yeah I would love some design input here, I have tried 2 or 3 iterations locally that I all hated. 😬 
cc @danielmarquespt when you get to it.

courier-new commented on client/web/src/enterprise/executors/instances/ExecutorsListPage.tsx:
```diff
@@ -0,0 +1,119 @@
+                </Text>
+            </Container>
+            <Container className="mb-3">
+                <FilteredConnection<ExecutorFields>
```
Nit: we're inconsistent between this and the secrets list in terms of centered vs. left-aligned summary and `noSummaryIfAllNodesVisible`.

courier-new commented on client/web/src/enterprise/executors/secrets/RemoveSecretModal.tsx:
```diff
@@ -0,0 +1,67 @@
+
+    return (
+        <Modal onDismiss={onCancel} aria-labelledby={labelId}>
+            <H3 id={labelId}>Remove secret {secret.key}?</H3>
```
```suggestion
            <H3 id={labelId}>Executor secret: {secret.key}</H3>
```

Might be more our style, looking at the Remove Credentials modal for inspiration.

courier-new commented on client/web/src/enterprise/executors/secrets/SecretAccessLogsModal.tsx:
```diff
@@ -0,0 +1,91 @@
+
+    return (
+        <Modal onDismiss={onCancel} aria-labelledby={labelId}>
+            <H3 id={labelId}>Access log</H3>
```
```suggestion
            <H3 id={labelId}>Access logs</H3>
```

courier-new commented on client/web/src/enterprise/executors/secrets/SecretAccessLogsModal.tsx:
```diff
@@ -0,0 +1,91 @@
+    return (
+        <Modal onDismiss={onCancel} aria-labelledby={labelId}>
+            <H3 id={labelId}>Access log</H3>
+            <ConnectionContainer className="mb-3">
```
Like we do with most of our other connections, maybe we should preface this with a 1-sentence text summary of what this list represents.

courier-new commented on client/web/src/enterprise/executors/secrets/backend.ts:
```diff
@@ -0,0 +1,287 @@
+            __typename
+            ... on ExecutorSecret {
+                accessLogs(first: $first, after: $after) {
+                    ...ExecutorSecretAccessLogConnectionFields
```
```suggestion
                    ...ExecutorSecretAccessLogsConnectionFields
```

courier-new commented on client/web/src/enterprise/executors/secrets/backend.ts:
```diff
@@ -0,0 +1,287 @@
+        }
+    }
+
+    fragment ExecutorSecretAccessLogConnectionFields on ExecutorSecretAccessLogConnection {
```
```suggestion
    fragment ExecutorSecretAccessLogsConnectionFields on ExecutorSecretAccessLogConnection {
```

I think the GraphQL schema type should also be plural but I know that was from the last PR, sorry. 😬

courier-new commented on client/web/src/enterprise/executors/secrets/backend.ts:
```diff
@@ -0,0 +1,287 @@
+    ${EXECUTOR_SECRET_CONNECTION_FIELDS}
+`
+
+export const userExecutorSecretsConnectionFactory = (
+    user: Scalars['ID'],
+    scope: ExecutorSecretScope
+): UseConnectionResult<ExecutorSecretFields> =>
+    // Scope has to be injected dynamically.
+    // eslint-disable-next-line react-hooks/rules-of-hooks
```
I feel like I'm missing something. Why does this seem to need a different treatment from `useExecutorSecretAccessLogsConnection` down below? That one injects the variable `secret` the exact same way, no?  I would have expected this to just be defined like:

```suggestion
export const useUserExecutorSecretsConnection = (
    user: Scalars['ID'],
    scope: ExecutorSecretScope
): UseConnectionResult<ExecutorSecretFields> =>
```

I think the linter warning is just over a hook being called from a non-component, non-hook-function, right? I think just naming it in the "hook-y" way (starts with `use`) is all that's going on here.

eseliger replied:
Yes but then it complains about the use in `useCallback` above, which is required because the child list component injects the scope to the parent using that. 
It's a bit weirdly coupled sadly.

courier-new commented on client/web/src/enterprise/executors/secrets/backend.ts:
```diff
@@ -0,0 +1,287 @@
+    ${EXECUTOR_SECRET_CONNECTION_FIELDS}
+`
+
+export const orgExecutorSecretsConnectionFactory = (
+    org: Scalars['ID'],
+    scope: ExecutorSecretScope
+): UseConnectionResult<ExecutorSecretFields> =>
+    // Scope has to be injected dynamically.
+    // eslint-disable-next-line react-hooks/rules-of-hooks
```
(same)

courier-new commented on client/web/src/enterprise/executors/secrets/backend.ts:
```diff
@@ -0,0 +1,287 @@
+    ${EXECUTOR_SECRET_CONNECTION_FIELDS}
+`
+
+export const globalExecutorSecretsConnectionFactory = (
+    scope: ExecutorSecretScope
+): UseConnectionResult<ExecutorSecretFields> =>
+    // Scope has to be injected dynamically.
+    // eslint-disable-next-line react-hooks/rules-of-hooks
```
(and again)

courier-new commented on internal/database/executor_secrets.go:
```diff
@@ -53,7 +55,7 @@ func (e ExecutorSecret) Value(ctx context.Context, s ExecutorSecretAccessLogStor
 
 const (
-	ExecutorSecretScopeBatches = "batches"
+	ExecutorSecretScopeBatches ExecutorSecretScope = "batches"
```
Again with the "batches" vs. "batch changes" 😭

eseliger replied:
this is meant to match the executor queue name this is meant to be used for, so I think it makes sense to name it batches here

courier-new reviewed (approved):
In general looks really great! 😄 Mostly left a bunch of nits and tried to make some design suggestions, but I'm no replacement for Daniel. 😛

eseliger commented on client/web/src/enterprise/batches/batch-spec/execute/workspaces/WorkspaceDetails.tsx:
```diff
@@ -611,7 +611,8 @@ const WorkspaceStep: React.FunctionComponent<React.PropsWithChildren<WorkspaceSt
                                             {step.environment.map(variable => (
                                                 <li key={variable.name}>
-                                                    {variable.name}: {variable.value}
+                                                    {variable.name}: {variable.value !== null && <>{variable.value}</>}
+                                                    {variable.value === null && <i>Set from secret</i>}
```
I actually find 2 separate statements a bit more readable 😬 Ternaries are sometimes a bit hard to track in the head. No strong feelings though

eseliger commented on client/web/src/enterprise/executors/instances/ExecutorCompatibilityAlert.story.tsx:
```diff
@@ -0,0 +1,58 @@
+
+export default config
+
+export const UpToDate: Story = () => (
+    <WebStory>
+        {props => (
+            <ExecutorCompatibilityAlert
+                {...props}
+                compatibility={ExecutorCompatibility.UP_TO_DATE}
+                hostname="executor1.sgdev.org"
+            />
+        )}
+    </WebStory>
+)
```
My thinking was that it would (if we ever get screenshot testing again) catch regressions that this doesn't render anything

eseliger replied:
added the comment!

eseliger commented on client/web/src/enterprise/executors/secrets/AddSecretModal.story.tsx:
```diff
@@ -0,0 +1,38 @@
+    decorators: [decorator],
+    parameters: {
+        chromatic: {
+            // Delay screenshot taking, so the modal has opened by the time the screenshot is taken.
+            delay: 2000,
```
weirdly no, all our modals had that problem in the past, so just copy pasted it haha.

eseliger commented:
Addressed most of the comments and pinged design on the others. 

@danielmarquespt This feature should get a proper design QA, so that we can close this chapter once and for all. 
I don't think this is critical for merging this PR though, so I am going ahead with that soon, to validate that things work end to end in production, where we run firecracker executors. Don't want to have a surprise moment next week close before branch cut :)
